### PR TITLE
Thick sample

### PR DESCRIPTION
**Pull request by isildur7**
Added a multislice thick sample

Added a method for precalculating propagators outside

Should we modify the exact_propagation functions to be able to take a propagator?

### CODE REVIEW

File: src/chromatix/functional/propagation.py [hunk 1]
```diff
@@ -161,6 +161,47 @@ def exact_propagate(
...
+def calculate_exact_propagator(
+    shape: Tuple,
+    dx: Union[float, Array],
+    spectrum,
+    z: float,
+    n: float,
+    N_pad: int,
+    kykx: Array = jnp.zeros((2,)),
+):
+    """Calculate a proapgator which can be used in the exact propagate method.
```
Comment (diptodip) [inline src/chromatix/functional/propagation.py:173]: proapgator -> propagator

File: src/chromatix/functional/samples.py [hunk 2]
```diff
@@ -30,14 +37,91 @@ def thin_sample(field: Field, absorption: Array, dn: Array, thickness: Array) ->
...
+            field.u.shape, field.dx, field.spectrum, thickness_per_slice, n, N_pad, kykx
+        )
+
+    def _update_per_slice(i: int, field: Field):
+        # Propagate the field
+        absorption = (absorption_stack[i])[jnp.newaxis, :, :, jnp.newaxis]
+        dn = (dn_stack[i])[jnp.newaxis, :, :, jnp.newaxis]
+
+        field = thin_sample(field, absorption, dn, thickness_per_slice)
+        # Propagating field
```
Comment (diptodip) [inline src/chromatix/functional/samples.py:102]: Could we pass the calculated propagator to exact_propagate here?

**Review by diptodip — CHANGES_REQUESTED** [review]
Could you implement passing the propagator as an option to exact_propagate?

Inside `exact_propagate` it should look something like 

```python
if propagator is None:
    propagator = calculate_exact_propagator(...)
```
so that the current behavior stays the same if we don't pass the propagator.

And then in the `Propagate` element, we can precompute the propagator during setup and pass it to the propagation function as needed. So eventually we should do this for `transfer_propagate` as well.

I'm happy to make these changes myself.

**Comment by isildur7** [comment]
I will implement it for transfer_propagate as well, in another PR soon.

File: src/chromatix/functional/samples.py [hunk 3]
```diff
@@ -30,14 +37,93 @@ def thin_sample(field: Field, absorption: Array, dn: Array, thickness: Array) ->
...
+            propagator=propagator,
+            mode="same",
+        )
+
+    # python loop unrolling is the fastest method
+    for i in range(absorption_stack.shape[0]):
+        field = _update_per_slice(i, field)
+
+    # propagate field backwards to the middle
+    half_stack_thikness = thickness_per_slice * absorption_stack.shape[0] / 2
```
Comment (diptodip) [inline src/chromatix/functional/samples.py:118]: `half_stack_thikness -> half_stack_thickness`

File: src/chromatix/functional/samples.py [hunk 4]
```diff
@@ -30,14 +37,93 @@ def thin_sample(field: Field, absorption: Array, dn: Array, thickness: Array) ->
...
+
+    field = exact_propagate(
+        field,
+        z=-half_stack_thikness,
```
Comment (diptodip) [inline src/chromatix/functional/samples.py:122]: And here

**Review by diptodip — COMMENTED** [review]
So of course it mostly looks good, we just have to nail down exactly how we want the custom propagator arguments handled in https://github.com/TuragaLab/chromatix/discussions/78

**Comment by diptodip** [comment]
But once those little typos are fixed I'm happy to accept this one and then fix the custom propagator stuff in another PR.

**Review by diptodip — APPROVED** [review]

**Pull request merged**